### ISSUE TITLE
Ceedling options issue..

### ISSUE THREAD

**Issue by krish2487** [ThrowTheSwitch/Ceedling#589]
Hello, 
I stumbled onto a very interesting thing while trying to redo the exercises of the second course over at udemy. 
I pulled the docker image and made a folder and setup the course quite a while ago.. maybe an year plus.... All well and good.. I was trying to recollect how I tested the different iterations earlier and spent some time searching around.. and I sort of eventually got there..

However, this is where the behavior gets interesting.
When I run 'ceedling options:*' in the Logger folder, I get this error.

```
root@97f1866c8419:/lab/Logger# ceedling options:*
ERROR: Found no file '*.yml' in search paths.
rake aborted!

/var/lib/gems/2.4.0/gems/ceedling-0.30.0/lib/ceedling/file_finder_helper.rb:45:in `blow_up'
/var/lib/gems/2.4.0/gems/ceedling-0.30.0/lib/ceedling/file_finder_helper.rb:30:in `find_file_in_collection'
/var/lib/gems/2.4.0/gems/ceedling-0.30.0/lib/ceedling/file_finder.rb:146:in `find_file_from_list'
/var/lib/gems/2.4.0/gems/ceedling-0.30.0/lib/ceedling/tasks_base.rake:92:in `block (2 levels) in <top (required)>'
/var/lib/gems/2.4.0/gems/ceedling-0.30.0/bin/ceedling:346:in `block in <top (required)>'
/var/lib/gems/2.4.0/gems/ceedling-0.30.0/bin/ceedling:333:in `<top (required)>'
/usr/local/bin/ceedling:23:in `load'
/usr/local/bin/ceedling:23:in `<main>'
Tasks: TOP => options:*
(See full trace by running task with --trace)
ERROR: Ceedling Failed
root@97f1866c8419:/lab/Logger#
```
However, running 'ceedling options:iteration2 test:all' runs correctly.
This is what I see on the console

```
root@97f1866c8419:/lab/Logger# ceedling options:iteration2 test:all


Test 'TestAnalogConductor.c'
----------------------------
Running TestAnalogConductor.out...


Test 'TestAnalogHardware.c'
---------------------------
Running TestAnalogHardware.out...


Test 'TestAnalogModel.c'
------------------------
Running TestAnalogModel.out...


Test 'TestDigital.c'
--------------------
Running TestDigital.out...


Test 'TestExecutor.c'
---------------------
Running TestExecutor.out...


Test 'TestFilter.c'
-------------------
Running TestFilter.out...


Test 'TestLED.c'
----------------
Running TestLED.out...


Test 'TestMain.c'
-----------------
Running TestMain.out...


Test 'TestTimer.c'
------------------
Running TestTimer.out...


Test 'TestUtils.c'
------------------
Running TestUtils.out...

--------------------
IGNORED TEST SUMMARY
--------------------
[TestMain.c]
  Test: test_Main_NeedToImplement
  At line (15): "Main Not Implemented"

--------------------
OVERALL TEST SUMMARY
--------------------
TESTED:  48
PASSED:  47
FAILED:   0
IGNORED:  1
```

And I thought that it has been quite a while and maybe I made some change and forgot about it.. so I started from scratch and redid the same steps..
Created a new folder, ( after exiting the docker container of course.)
ran the docker command and mapped the volume correctly. 
Inside the container, ran the 'course setup' command and then entered the Logger folder.
Then ran the 'ceedling options:*' command and received the same error.
However, this time, when I ran the 'ceedling options:iteration2 test:all' The tests did not run successfully and gave this error.

```
root@069d8a45d2c8:/lab/Logger# ceedling options:iteration2 test:all


Test 'TestAnalogConductor.c'
----------------------------
Linking TestAnalogConductor.out...
build/test/out/c/MockAnalogModel.o: In function `MockAnalogModel_Destroy':
/lab/Logger/build/mocks/MockAnalogModel.c:95: undefined reference to `GlobalExpectCount'
/lab/Logger/build/mocks/MockAnalogModel.c:95: undefined reference to `GlobalVerifyOrder'
build/test/out/c/MockAnalogModel.o: In function `AnalogModel_Init':
/lab/Logger/build/mocks/MockAnalogModel.c:116: undefined reference to `GlobalVerifyOrder'
build/test/out/c/MockAnalogModel.o: In function `AnalogModel_Init_CMockExpect':
/lab/Logger/build/mocks/MockAnalogModel.c:138: undefined reference to `GlobalExpectCount'
build/test/out/c/MockAnalogModel.o: In function `AnalogModel_AddReading':
/lab/Logger/build/mocks/MockAnalogModel.c:167: undefined reference to `GlobalVerifyOrder'
build/test/out/c/MockAnalogModel.o: In function `AnalogModel_AddReading_CMockExpect':
/lab/Logger/build/mocks/MockAnalogModel.c:196: undefined reference to `GlobalExpectCount'
build/test/out/c/MockAnalogModel.o: In function `AnalogModel_GetChannel':
/lab/Logger/build/mocks/MockAnalogModel.c:225: undefined reference to `GlobalVerifyOrder'
build/test/out/c/MockAnalogModel.o: In function `AnalogModel_GetChannel_CMockExpectAndReturn':
/lab/Logger/build/mocks/MockAnalogModel.c:264: undefined reference to `GlobalExpectCount'
build/test/out/c/MockAnalogHardware.o: In function `MockAnalogHardware_Destroy':
/lab/Logger/build/mocks/MockAnalogHardware.c:115: undefined reference to `GlobalExpectCount'
/lab/Logger/build/mocks/MockAnalogHardware.c:115: undefined reference to `GlobalVerifyOrder'
build/test/out/c/MockAnalogHardware.o: In function `AnalogHardware_Init':
/lab/Logger/build/mocks/MockAnalogHardware.c:136: undefined reference to `GlobalVerifyOrder'
build/test/out/c/MockAnalogHardware.o: In function `AnalogHardware_Init_CMockExpect':
/lab/Logger/build/mocks/MockAnalogHardware.c:158: undefined reference to `GlobalExpectCount'
build/test/out/c/MockAnalogHardware.o: In function `AnalogHardware_SetChans':
/lab/Logger/build/mocks/MockAnalogHardware.c:183: undefined reference to `GlobalVerifyOrder'
build/test/out/c/MockAnalogHardware.o: In function `AnalogHardware_SetChans_CMockExpect':
/lab/Logger/build/mocks/MockAnalogHardware.c:211: undefined reference to `GlobalExpectCount'
build/test/out/c/MockAnalogHardware.o: In function `AnalogHardware_IsReady':
/lab/Logger/build/mocks/MockAnalogHardware.c:241: undefined reference to `GlobalVerifyOrder'
build/test/out/c/MockAnalogHardware.o: In function `AnalogHardware_IsReady_CMockExpectAndReturn':
/lab/Logger/build/mocks/MockAnalogHardware.c:281: undefined reference to `GlobalExpectCount'
build/test/out/c/MockAnalogHardware.o: In function `AnalogHardware_GetReading':
/lab/Logger/build/mocks/MockAnalogHardware.c:310: undefined reference to `GlobalVerifyOrder'
build/test/out/c/MockAnalogHardware.o: In function `AnalogHardware_GetReading_CMockExpectAndReturn':
/lab/Logger/build/mocks/MockAnalogHardware.c:349: undefined reference to `GlobalExpectCount'
collect2: error: ld returned 1 exit status
ERROR: Shell command failed.
> Shell executed command:
'arm-none-eabi-gcc -mcpu=cortex-m3 -mthumb -T test/support/qemu.ld build/test/out/c/TestAnalogConductor_runner.o build/test/out/c/TestAnalogConductor.o build/test/out/c/MockAnalogModel.o build/test/out/c/MockAnalogHardware.o build/test/out/c/unity.o build/test/out/c/AnalogConductor.o build/test/out/c/startup.o build/test/out/c/syscalls.o build/test/out/c/uart.o build/test/out/c/cmock.o -o build/test/out/TestAnalogConductor.out -lm'
> And exited with status: [1].


NOTICE: If the linker reports missing symbols, the following may be to blame:
  1. Test lacks #include statements corresponding to needed source files.
  2. Project search paths do not contain source files corresponding to #include statements in the test.
  3. Test does not #include needed mocks.

rake aborted!

/var/lib/gems/2.4.0/gems/ceedling-0.30.0/lib/ceedling/generator.rb:151:in `rescue in generate_executable_file'
/var/lib/gems/2.4.0/gems/ceedling-0.30.0/lib/ceedling/generator.rb:155:in `generate_executable_file'
/var/lib/gems/2.4.0/gems/ceedling-0.30.0/lib/ceedling/rules_tests.rake:39:in `block in <top (required)>'
/var/lib/gems/2.4.0/gems/ceedling-0.30.0/lib/ceedling/task_invoker.rb:107:in `invoke_test_results'
/var/lib/gems/2.4.0/gems/ceedling-0.30.0/lib/ceedling/test_invoker.rb:124:in `block in setup_and_invoke'
/var/lib/gems/2.4.0/gems/ceedling-0.30.0/lib/ceedling/test_invoker.rb:51:in `setup_and_invoke'
/var/lib/gems/2.4.0/gems/ceedling-0.30.0/lib/ceedling/tasks_tests.rake:11:in `block (2 levels) in <top (required)>'
/var/lib/gems/2.4.0/gems/ceedling-0.30.0/bin/ceedling:346:in `block in <top (required)>'
/var/lib/gems/2.4.0/gems/ceedling-0.30.0/bin/ceedling:333:in `<top (required)>'
/usr/local/bin/ceedling:23:in `load'
/usr/local/bin/ceedling:23:in `<main>'

Caused by:
ShellExecutionException: ShellExecutionException
/var/lib/gems/2.4.0/gems/ceedling-0.30.0/lib/ceedling/tool_executor.rb:88:in `exec'
/var/lib/gems/2.4.0/gems/ceedling-0.30.0/lib/ceedling/generator.rb:136:in `generate_executable_file'
/var/lib/gems/2.4.0/gems/ceedling-0.30.0/lib/ceedling/rules_tests.rake:39:in `block in <top (required)>'
/var/lib/gems/2.4.0/gems/ceedling-0.30.0/lib/ceedling/task_invoker.rb:107:in `invoke_test_results'
/var/lib/gems/2.4.0/gems/ceedling-0.30.0/lib/ceedling/test_invoker.rb:124:in `block in setup_and_invoke'
/var/lib/gems/2.4.0/gems/ceedling-0.30.0/lib/ceedling/test_invoker.rb:51:in `setup_and_invoke'
/var/lib/gems/2.4.0/gems/ceedling-0.30.0/lib/ceedling/tasks_tests.rake:11:in `block (2 levels) in <top (required)>'
/var/lib/gems/2.4.0/gems/ceedling-0.30.0/bin/ceedling:346:in `block in <top (required)>'
/var/lib/gems/2.4.0/gems/ceedling-0.30.0/bin/ceedling:333:in `<top (required)>'
/usr/local/bin/ceedling:23:in `load'
/usr/local/bin/ceedling:23:in `<main>'
Tasks: TOP => build/test/results/TestAnalogConductor.pass => build/test/out/TestAnalogConductor.out
(See full trace by running task with --trace)
ERROR: Ceedling Failed
```

Has something changed or am I doing something wrong??
I pulled the latest image from docker hub before attempting to run the commands.. so both the folders must be running from the same docker image anyway.. 

Thanks in advance!!!

**Comment by Letme** [comment]
Did you handle permissions (owners) as well? Although error is strange, I would also assume it is timestamp based

**Comment by krish2487** [comment]
I am working on a Mac, I have not needed to explicitly set any permissions or remember having needing to tinker with it earlier.
But I am game if you do have some suggestions on what I can try..

Anything I can do to make it a reproducible issue and help isolate the cause?

Thanks again!! 😄

**Comment by Letme** [comment]
I would start with clean directory every time before you run docker. That means do not have a cascading build, but rather a fresh mapping without exporting artifacts back.

I don't know mac, but there is --user command to docker so that you do not get root owned files in your drive. What then happens is that files are readable and there, but cannot be overwritten which leads to strange leftovers of "old" object files.

**Comment by krish2487** [comment]
I did.. I created a separate folder and tried the commands afresh. That is the behavior I was reporting.
The error while running the "ceedling options:iteration2 test:all" is what I observed when trying to run the commands in a new, empty directory.

When i reran the commands on the old, existing repository then I was able to run the tests.

In both the cases "ceedling options:*" returns the same error.

**Comment by Letme** [comment]
But I assume you confirmed that `/lab/Logger` contains `project.yml` ?

**Comment by krish2487** [comment]
Umm.. sorry.. Let me have a do over..
When I did the second course at udemy, I created a folder ( lets call it folderA ) on my system, pulled the docker image.. ran the requisite commands and finished the course.. that's done. 

I wanted to refresh my memory about the "options_path" and did a docker pull and tried to rerun the "ceedling options:*" and the "ceedling options:iteration2 test:all" in the same folderA. Now the "options:*" failed with the no file found message.. while the tests ran.

To verify that I did not change anything on the folderA and forgot about it.. I created another folderB on the same directory level as folderA. I entered the folder, executed the docker command and upon entering the container, executed the "course setup" which created the requisite files and directories.. I then re-ran the "ceedling options:*" and "ceedling options:iteration2 test:all" commands.. except in this case, both the commands failed and gave an error. 

All this while I was in folderB.

I hope this is a lot more clearer than the obvious mess I made while trying to explain earlier.. :-P 

Both the folders - folderA and folderB have the same contents.. and yet exhibit different behaviors.. the only difference is that the folderA was created quite a while ago.. while I was taking the course.. folderB is what I have created today..

**Comment by mvandervoord** [comment]
The option `ceedling options:*` isn't really an option. It's a notation used to show that you can enter any of the option file names after options: and get the option you want. You'll see the same notation used for running individual tests. To clear up this confusion, the latest version of ceedling lists the individual options (as there likely aren't many).

Given this information, the first error becomes more clear. When you specify `ceedling options:* test:all` it attempts to find a file named `*.yml` in your options folder and doesn't find one. When you issue `ceedling options:iteration1 test:all` it CAN find the option `iteration1.yml` in the options folder, and so it runs just fine.

For the second error, I suspect that if you do a diff between your two project.yml files (the MAIN project.yml file, not the options files) you will see that they have a differences in global ordering (and possibly other cmock features). It's working in the first case because you would have been keeping up with these changes as the course progressed. When you check out the latest version, it does NOT have these changes to the main project.yml and therefore it's not quite fully able to run them.

**Comment by krish2487** [comment]
@mvandervoord  
Thanks.. I assumed the "ceedling options:\*" would list all the options available.. My mistake.. Thank you for clarifying that.
It is possible that the docker image might be a version or two behind.

Spot on with the second question..
I did a diff between the different project.yml files and I see that several cmock options are commented out in the newer project folder.

```
 diff tdd/Logger/project.yml throwtheswitch_2/Logger/project.yml
72,76c72,76
<     - :array
<     - :expect_any_args
<     - :return_thru_ptr
<     - :ignore_arg
<     - :callback
---
>     #- :array
>     #- :expect_any_args
>     #- :return_thru_ptr
>     #- :ignore_arg
>     #- :callback
```
To be clear , in the code block above.. the "tdd" folder is the old project folder that I created at the time of doing the course.. and the "throwtheswitch_2" is what I created a couple of days ago.

However, I did uncomment those options and reran "ceedling options:iteration2 test:all" and still see the same errors.. That made me check if the iteration2.yml file was different between the different project folders and I ran a diff to check the differences if any.

```
15a16,35
> :cmock:
>   :mock_path: 'build/mocks/'
>   :includes:
>     - stdint.h
>     - Defs.h
>     - LPC1768.h
>   :includes_h_pre_orig_header:
>     - stdint.h
>     - Defs.h
>     - LPC1768.h
>   :enforce_strict_ordering: TRUE
>   :verbosity: 1
>   :plugins:
>     - :expect
>     - :ignore
>     #- :array
>     #- :expect_any_args
>     #- :return_thru_ptr
>     #- :ignore_arg
>     #- :callback

```

The new iteration2.yml file has quite a few cmock options defined in it, that for some reason do not allow the tests to compile. Removing the cmock section from the iteration2.yml file in the new project folder allowed the tests to run and compile successfully like the old project folder. Does it work for you guys??  

So thanks again @mvandervoord  and @Letme , it is always a positively wonderful experience to ask for help and be guided to the correct answer here!! 😄 .

I am closing the issue now..
